api-sdk/src/main/java/com/smartling/api/sdk/exceptions/ApiException.java

```diff
@@ -23,13 +34,68 @@
 {
     private static final long serialVersionUID = -397098626101615761L;
 
-    public ApiException(final String message)
+    private int httpCode;
+    private List<String> messages = new ArrayList<>();
+
+    ApiException(final String contents, List<String> messages, int httpCode)
     {
-        super(message);
+        super(contents);
+        this.messages = messages;
+        this.httpCode = httpCode;
     }
 
-    public ApiException(final Exception e)
+    ApiException(final Exception e)
     {
         super(e);
+        messages.add(e.getMessage());
+    }
+
+    public static ApiException newException(String contents, int httpCode)
+    {
+        ApiResponse<EmptyResponse> apiResponse = JsonReader.parseApiResponse(contents, new TypeToken<ApiResponseWrapper<EmptyResponse>>()
+                {
+                }
+        );
+        String apiCode = apiResponse.getCode();
+        List<String> messages = apiResponse.getMessages();
+        return getApiException(contents, httpCode, apiCode, messages);
+    }
+
+    private static ApiException getApiException(final String contents, final int httpCode, final String apiCode, final List<String> messages)
+    {
+        switch (apiCode)
+        {
+            case "VALIDATION_ERROR":
+                return new ValidationException(contents, messages, httpCode);
+            case "AUTHENTICATION_ERROR":
+                return new AuthenticationException(contents, messages, httpCode);
+            case "AUTHORIZATION_ERROR":
+                return new AuthorizationException(contents, messages, httpCode);
+            case "RESOURCE_LOCKED":
+                return new ResourceLockedException(contents, messages, httpCode);
+            case "MAX_OPERATIONS_LIMIT_EXCEEDED":
+                return new OperationsLimitExceeded(contents, messages, httpCode);
+            case "GENERAL_ERROR":
+                return new UnexpectedException(contents, messages, httpCode);
+            case "MAINTENANCE_MODE_ERROR":
+                return new ServiceTemporaryUnavailableException(contents, messages, httpCode);
+            default:
+                return new ApiException(contents, messages, httpCode);
+        }
```

Comment:
I would think of some exception hierarchy here. For instance, ResourceLocked, OperationLimitsExceeded, Maintenance mean pretty much the same for the client: "not available now, try later". So the likely to be handled in the similar fashion. Maybe extract a common superclass for them ?